compiler/next/include/chpl/resolution/resolution-types.h

```diff
@@ -35,105 +39,553 @@ namespace resolution {
+    return PoiInfo::reuseEquals(*this, other);
+  }
+  bool operator!=(const PoiInfo& other) const {
+    return !(*this == other);
+  }
 };
 
+// TODO: should this actually be types::FunctionType?
 struct TypedFnSignature {
-  UniqueString name;
-  bool isMethod; // in that case, formals[0] is the receiver
-  std::vector<TypedFnSignatureFormal> formals;
+  typedef enum {
+    WHERE_NONE,  // no where clause
+    WHERE_TBD,   // where clause not resolved yet
+    WHERE_TRUE,  // where resulted in true
+    WHERE_FALSE, // where resulted in false
+  } WhereClauseResult;
+
+  // What is the untyped function signature?
+  const UntypedFnSignature* untypedSignature;
+  // What is the type of each of the formals?
+  std::vector<types::QualifiedType> formalTypes;
+  // If there was a where clause, what was the result of evaluating it?
+  WhereClauseResult whereClauseResult = WHERE_TBD;
+
+  // Are any of the formals generic or unknown at this point?
+  bool needsInstantiation = true;
+
+  // Is this TypedFnSignature representing an instantiation?
+  // If so, what is the generic TypedFnSignature that was instantiated?
+  const TypedFnSignature* instantiatedFrom = nullptr;
+
+  // Is this for an inner Function? If so, what is the parent
+  // function signature?
+  const TypedFnSignature* parentFn = nullptr;
+
+  // TODO: This could include a substitutions map, if we need it.
+  // The formalTypes above might be enough, though.
+
+  bool operator==(const TypedFnSignature& other) const {
+    return untypedSignature == other.untypedSignature &&
+           formalTypes == other.formalTypes &&
+           whereClauseResult == other.whereClauseResult &&
+           needsInstantiation == other.needsInstantiation &&
+           instantiatedFrom == other.instantiatedFrom &&
+           parentFn == other.parentFn;
+  }
+  bool operator!=(const TypedFnSignature& other) const {
+    return !(*this == other);
+  }
+
+  std::string toString() const;
+
+  const ID& functionId() const {
+    return untypedSignature->functionId;
+  }
+};
+
+class MostSpecificCandidates {
+ public:
+  typedef enum {
+    REF = 0,
+    CONST_REF = 1,
+    VALUE = 2,
+    N = 3,
+  } Intent;
+
+ private:
+  const TypedFnSignature* candidates[N] = {nullptr};
+
+ public:
+  const TypedFnSignature* const* begin() const {
+    return &candidates[0];
+  }
+  const TypedFnSignature* const* end() const {
+    return &candidates[N];
+  }
+
+  void setBestRef(const TypedFnSignature* sig) {
+    candidates[REF] = sig;
+  }
+  void setBestConstRef(const TypedFnSignature* sig) {
+    candidates[CONST_REF] = sig;
+  }
+  void setBestValue(const TypedFnSignature* sig) {
+    candidates[VALUE] = sig;
+  }
+
+  const TypedFnSignature* bestRef() const {
+    return candidates[REF];
+  }
+  const TypedFnSignature* bestConstRef() const {
+    return candidates[CONST_REF];
+  }
+  const TypedFnSignature* bestValue() const {
+    return candidates[VALUE];
+  }
+
+  const TypedFnSignature* only() const {
```

Comment:
Can you say more about why you think this one should have `candidate` in the name (and presumably you don't feel that way about the other methods here)? I'm open to it but I don't yet see the reason to do so. The type `MostSpecificCandidates` just contains candidates... Are you thinking that one might suppose that `only()` returns a `bool` or something? Or refers to this being the only MostSpecificCandidates instance somehow?

Reply:
I think it's probably because I'm so used to "only" for "only lists" in use statements.  Which is probably a personal association that most other compiler developers wouldn't have.

Reply:
Yeah, that didn't even occur to me :) But I think the context here is sufficiently different.